src/mirrors.rs

```diff
@@ -0,0 +1,240 @@
+/// A mirrored PostgreSQL client.
+/// Packets arrive to us through a channel from the main client and we send them to the server.
+use std::cmp::{max, min};
+use std::time::Duration;
+
+use bytes::{Bytes, BytesMut};
+
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::time::sleep;
+
+use crate::config::{Address, Role, User};
+use crate::errors::Error;
+use crate::messages::flush;
+use crate::pool::ClientServerMap;
+use crate::server::Server;
+use crate::stats::get_reporter;
+use log::{error, info, trace, warn};
+
+const MAX_CONNECT_RETRIES: u32 = 5;
+const MAX_SEND_RETRIES: u32 = 3;
+
+pub struct MirroredClient {
+    address: Address,
+    user: User,
+    database: String,
+    bytes_rx: Receiver<Bytes>,
+    disconnect_rx: Receiver<()>,
+    successful_sends_without_recv: u32,
+}
+
+impl MirroredClient {
+    async fn recv_if_neccessary(&mut self, server: &mut Server) -> Result<(), Error> {
+        /* We only receive a response from the server if we have successfully sent
+          5 messages on the current connection. We also send a flush message to gaurantee
+          a server response.
+        */
+        if self.successful_sends_without_recv >= 5 {
+            // We send a flush message to gaurantee a server response
+            server.send(&flush()).await?;
+            server.recv().await?;
+            self.successful_sends_without_recv = 0;
+            trace!(
+                "Received a response from mirror server {:?}",
+                server.address()
+            );
+        }
+        Ok(())
+    }
+
+    async fn connect_with_retries(&mut self, server_id: i32) -> Option<Server> {
+        let mut delay = Duration::from_secs(0);
+        let min_backoff = Duration::from_millis(100);
+        let max_backoff = Duration::from_secs(5);
+        let mut retries = 0;
+
+        loop {
+            if let Some(tmp_server) = self.connect(server_id).await {
+                if !tmp_server.is_bad() {
+                    break Some(tmp_server);
+                }
+            }
+            delay = max(min_backoff, delay);
+            delay = min(max_backoff, delay * 2);
+            retries += 1;
+            if retries > MAX_CONNECT_RETRIES {
+                break None;
+            }
+            sleep(delay).await;
+        }
+    }
+
+    async fn connect(&mut self, server_id: i32) -> Option<Server> {
+        self.successful_sends_without_recv = 0;
+        let stats = get_reporter();
+        stats.server_register(
+            server_id,
+            self.address.id,
+            self.address.name(),
+            self.database.clone(),
+            self.address.username.clone(),
+        );
+        stats.server_login(server_id);
+
+        match Server::startup(
```

Comment:
Not using bb8?

Reply:
Right. But to be honest, I hate the connection management part of this PR to the point I am okay to create a bb8 pool with one connection to not have to manage the lifetime of the connection.

Reply:
I decided to go with the bb8 route even if the pool is tiny to avoid having to handle connection lifetime